diff --git a/quantum/process_keycode/process_tap_dance.h b/quantum/process_keycode/process_tap_dance.h
index ca12f4746..15188c842 100644
--- a/quantum/process_keycode/process_tap_dance.h
+++ b/quantum/process_keycode/process_tap_dance.h
@@ -67,6 +67,12 @@ typedef struct
     .user_data = (void *)&((qk_tap_dance_pair_t) { kc1, kc2 }),  \
   }
 
+#define ACTION_TAP_DANCE_DOUBLE_TIME(kc1, kc2, tap_specific_tapping_term) { \
+    .fn = { qk_tap_dance_pair_on_each_tap, qk_tap_dance_pair_finished, qk_tap_dance_pair_reset }, \
+    .user_data = (void *)&((qk_tap_dance_pair_t) { kc1, kc2 }),  \
+    .custom_tapping_term = tap_specific_tapping_term, \
+  }
+
 #define ACTION_TAP_DANCE_DUAL_ROLE(kc, layer) { \
     .fn = { qk_tap_dance_dual_role_on_each_tap, qk_tap_dance_dual_role_finished, qk_tap_dance_dual_role_reset }, \
     .user_data = (void *)&((qk_tap_dance_dual_role_t) { kc, layer }), \
